### PR TITLE
Environment Linking/Unlinking for API

### LINKED ISSUE #193: Environment Linking - API
Create environment linking endpoints in the layer0 api

### PR DESCRIPTION
**Pull request by zpatrick**
closes #193 

This still needs unit tests, which will be added in a separate PR

### CODE REVIEW

File: api/backend/ecs/environment_manager.go [hunk 1]
```diff
@@ -327,6 +328,102 @@ func (this *ECSEnvironmentManager) DeleteEnvironment(environmentID string) error
...
+	sourceECSID := id.L0EnvironmentID(sourceEnvironmentID).ECSEnvironmentID()
+	destECSID := id.L0EnvironmentID(destEnvironmentID).ECSEnvironmentID()
+
+	sourceGroup, err := this.EC2.DescribeSecurityGroup(sourceECSID.SecurityGroupName())
+	if err != nil {
+		return err
+	}
+
+	if sourceGroup == nil {
+		log.Warnf("Skipping environment unlink since security group '%s' does not exist", sourceECSID.SecurityGroupName())
```
Comment (diemonster) [inline api/backend/ecs/environment_manager.go]: Maybe a philosophical question, but wouldn't this be an error? Most delete operations I can think of return an error if the thing being deleted doesn't exist.

File: api/backend/ecs/environment_manager.go [hunk 2]
```diff
@@ -327,6 +328,102 @@ func (this *ECSEnvironmentManager) DeleteEnvironment(environmentID string) error
 	return nil
 }
 
+func (this *ECSEnvironmentManager) CreateEnvironmentLink(sourceEnvironmentID, destEnvironmentID string) error {
```
Comment (diemonster) [inline api/backend/ecs/environment_manager.go]: still using `this`

File: api/backend/ecs/environment_manager.go [hunk 3]
```diff
@@ -327,6 +328,102 @@ func (this *ECSEnvironmentManager) DeleteEnvironment(environmentID string) error
...
+	if err := this.EC2.AuthorizeSecurityGroupIngressFromGroup(destGroup.GroupId, sourceGroup.GroupId); err != nil {
+		if !ContainsErrCode(err, "InvalidPermission.Duplicate") {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *ECSEnvironmentManager) DeleteEnvironmentLink(sourceEnvironmentID, destEnvironmentID string) error {
```
Comment (diemonster) [inline api/backend/ecs/environment_manager.go]: `this`

File: api/backend/ecs/environment_manager.go [hunk 4]
```diff
@@ -327,6 +328,102 @@ func (this *ECSEnvironmentManager) DeleteEnvironment(environmentID string) error
...
+}
+
+func (this *ECSEnvironmentManager) getEnvironmentSecurityGroup(environmentID id.ECSEnvironmentID) (*ec2.SecurityGroup, error) {
+	group, err := this.EC2.DescribeSecurityGroup(environmentID.SecurityGroupName())
+	if err != nil {
+		return nil, err
+	}
+
+	if group == nil {
+		return nil, fmt.Errorf("Security group for environment '%s' does not exist", environmentID.L0EnvironmentID())
```
Comment (diemonster) [inline api/backend/ecs/environment_manager.go]: slightly confused as to why this returns an error versus #307

Comment (zpatrick) [inline api/backend/ecs/environment_manager.go]: 307?

Comment (zpatrick) [inline api/backend/ecs/environment_manager.go]: this is a helper function used only in creating the link. It obviously can't create a link if the security group doesn't exist. However, delete actions need to be idempotent. If the security group doesn't exist, no unlinking work needs to be done, which satisfies the idempotency.

Comment (diemonster) [inline api/backend/ecs/environment_manager.go]: 👍

File: tests/system/environmentLink_test.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/quintilesims/layer0/tests/system/clients"
+	"testing"
+	"time"
+)
+
+// Test Resources:
+// This test creates two linked environments named 'el_public' and 'el_private`
+// The 'el_public' environment has a STS service running behind a public load balancer
+// The 'el_private' environment has a STS service running behind a private load balancer
+func TestEnvironmentLink(t *testing.T) {
```
Comment (diemonster) [inline tests/system/environmentLink_test.go]: [nit] consider adding an additional unlinked environment that attempts to connect to internal resources in either environment, or some variation of this idea

File: tests/system/windowsService_test.go [hunk 6]
```diff
@@ -11,6 +11,10 @@ import (
 // This test creates an environment named 'ws' that has a
 // Windows service named 'windows' running in iis
 func TestWindowsService(t *testing.T) {
+	if testing.Short() {
```
Comment (diemonster) [inline tests/system/windowsService_test.go]: Add an entry in the System Tests README.md for 'short' mode

File: tests/system/README.md [hunk 7]
```diff
@@ -65,7 +65,8 @@ When developing a system test, waiting for terraform to setup/destroy the test r
 Using this method, you can run your test multiple times without terraform destroying and rebuilding the resources.
 
 Some useful builtin flags:
-* `-run nameOfTest` - Executes tests that match the specified name (can be used to run a single test case)
-* `-parallel n` - Specifies the number of tests to run in parallel at once
+* `-run nameOfTest` - Executes tests that match the specified name (can be used to run a single test case).
+* `-parallel n` - Specifies the number of tests to run in parallel at once.
+* `-short` - Execute the tests in short mode. Long running tests will be skipped.
```
Comment (diemonster) [inline tests/system/README.md]: 👍  missed this the first time around.

File: tests/system/environmentLink_test.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+	publicServiceURL := s.Terraform.Output("public_service_url")
+	privateServiceURL := s.Terraform.Output("private_service_url")
+
+	publicService := clients.NewSTSTestClient(t, publicServiceURL)
+	publicService.WaitForHealthy(time.Minute * 3)
+
+	// curl the private service in the private environment from the public service in the public environment
+	// the private service returns "Hello, World!" from its root path
+	output, err := publicService.RunCommand("curl", "-s", privateServiceURL)
```
Comment (diemonster) [inline tests/system/environmentLink_test.go]: do we need to have `curl` here? Could Go stdlib work instead?
```
resp, err := http.Get(privateServiceURL)
if err != nil {
	// handle err
}
defer resp.Body.Close()
```

Comment (zpatrick) [inline tests/system/environmentLink_test.go]: [issue discussed person]

**Pull request merged**